clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/op/manifest/KubernetesPatchManifestOperation.java

```diff
@@ -115,7 +116,10 @@ private ReplaceResult replaceArtifacts(
         objectMapper.convertValue(description.getPatchBody(), KubernetesManifest.class);
     ReplaceResult replaceResult =
         patchHandler.replaceArtifacts(
-            manifest, allArtifacts, objToPatch.getNamespace(), description.getAccount());
+            manifest,
+            allArtifacts,
+            Strings.nullToEmpty(objToPatch.getNamespace()),
```

Comment:
This is really not that important but one argument for leaving the `namespaceMatches` input nullable is preserving symmetry with the nullability of the artifacts' namespace. Feels a little confusing to make callers `nullToEmpty` the namespace to match, but allow it to be null in an artifact, and might make it a bit clearer to have those two `nullToEmpty` conversions side-by-side in `namespaceMatches` to make it clear that we're going to match even if both namespaces are undefined.

Reply:
Yeah, I think that in cases where you're patching a cluster-scoped object you wouldn't need a namespace.  It actually looks like the UI still requires one but it's probably possible to omit it if you directly edit the JSON.

The way I've started thinking about things (at least in the Kubernetes provider is) that any  nullable `String` where null/empty are not semantically different should be canonicalized with `Strings.nullToEmpty` as soon as possible and marked non-nullable.  (With the main risk of that change being ensuring that we aren't doing any direct `== null` checks on that string anywhere.)

That does lead to asymmetry when comparing a `@Nullable` string with a `@Nonnull` string only requires a `nullToEmpty` on one of them.  At the location of the check that should be clear because we can see that one of them can't be null.  But I do see your point that in the call to `replaceArtifacts` we're passing in a bunch of artifacts with possibly-null locations, but a never-null namespace.  (Though I guess a counterpoint is that before this change it wasn't clear to the caller whether we'd match in the `null == ""` case.). I wish I could make `Artifact::getLocation` return an empty string instead of `null` but I think I tried that a while back and it was too risky.

I guess one further thing is that in an upcoming change I'll be making `objToPatch.getNamespace()` always non-null so we'll remove this `nullToEmpty` entirely (but the underlying issue you're bringing up will still be true, in that we'll be passing `""` as a namespace but `null` on the artifacts).

So I guess I'm sort of inclined to leave this, under the philosophy that unless there's a good reason, there's no distinction between `null` and `""` and you should just make them the same as soon as possible (which is here for the `namespace` but later for the result of `Artifact::getLocation`).  But let me know if this is confusing or you disagree!

Reply:
> any nullable String where null/empty are not semantically different should be canonicalized with Strings.nullToEmpty as soon as possible and marked non-nullable.

👍 👍  I'm on board w/ this philosophy!